File: rxcv4j/src/main/java/com/cv4j/rxjava/MemCache.java
package com.cv4j.rxjava;

import android.annotation.TargetApi;
import android.graphics.Bitmap;
import android.os.Build;
import android.support.v4.util.LruCache;

import com.safframwork.tony.common.utils.Preconditions;

/**
 * Created by Tony Shen on 2017/3/31.
 */

public class MemCache {

    private static final String TAG = "MemCache";

    private static LruCache<String, Bitmap> mLruCache; // 硬引用缓存

    private MemCache() {
        int cacheSize = (int) Runtime.getRuntime().maxMemory() / 8;
        mLruCache = new LruCache<String, Bitmap>(cacheSize) {
            @Override
            protected int sizeOf(String key, Bitmap value) {
                final int bitmapSize = getBitmapSize(value);
                return bitmapSize == 0 ? 1 : bitmapSize;
            }

            @Override
            protected void entryRemoved(boolean evicted, String key,
                                        Bitmap oldValue, Bitmap newValue) {
                if(evicted){
                    if(oldValue != null && !oldValue.isRecycled()){
                        oldValue = null;
                    }
                }
            }
        };
    }

    private static class SingletonHolder {
        private static final MemCache INSTANCE = new MemCache();
    }

    public static final MemCache getInstance() {
        return SingletonHolder.INSTANCE;
    }

    /**
     * @param value a bitmap
     * @return size of bitmap
     */
    @TargetApi(19)
    private int getBitmapSize(Bitmap value) {

        if (value == null) {
            return 0;
        }

        if (Build.VERSION.SDK_INT >= 19) {
            return value.getAllocationByteCount();
        }

        return value.getHeight() * value.getRowBytes();
    }

    /**
     * 清空lrucache缓存
     */
    public void clear() {
        mLruCache.evictAll();
    }

    /**
     * 根据key删除MemoryCache的图片缓存
     * @param key
     */
    public void remove(String key) {
        if (Preconditions.isNotBlank(key)) {
            mLruCache.remove(key);
        }
    }

    public synchronized void put(String key,Bitmap bitmap) {
        mLruCache.put(key, bitmap);
    }

    public Bitmap get(String key) {
        return mLruCache == null ? null : mLruCache.get(key);
    }
}

File: rxcv4j/src/main/java/com/cv4j/rxjava/MemCache.kt
/*
 * Copyright (c) 2017-present, CV4J Contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.cv4j.rxjava

import android.graphics.Bitmap
import android.os.Build
import android.support.v4.util.LruCache
import com.safframework.tony.common.utils.Preconditions

class MemCache private constructor() {

    init {
        val cacheSize = Runtime.getRuntime().maxMemory().toInt() / 8
        mLruCache = object : LruCache<String, Bitmap>(cacheSize) {
            override fun sizeOf(key: String?, value: Bitmap?): Int {
                val bitmapSize = getBitmapSize(value)
                return if (bitmapSize == 0) 1 else bitmapSize
            }

            override fun entryRemoved(evicted: Boolean, key: String?,
                                      oldValue: Bitmap?, newValue: Bitmap?) {

                if (evicted) {
                    if (oldValue != null && !oldValue.isRecycled) {
                        oldValue.recycle()
                    }
                }
            }
        }
    }

    private object SingletonHolder {
         val INSTANCE = MemCache()
    }

    /**
     * @param value a bitmap
     * *
     * @return size of bitmap
     */
    fun getBitmapSize(value: Bitmap?): Int {

        if (value == null) {
            return 0
        }

        if (Build.VERSION.SDK_INT >= 19) {
            return value.allocationByteCount
        } else {
            return value.height * value.rowBytes
        }
    }

    /**
     * 清空lrucache缓存
     */
    fun clear() {
        mLruCache!!.evictAll()
    }

    /**
     * 根据key删除MemoryCache的图片缓存
     * @param key
     */
    fun remove(key: String) {
        if (Preconditions.isNotBlank(key)) {
            mLruCache!!.remove(key)
        }
    }

    @Synchronized fun put(key: String?, bitmap: Bitmap) {
        mLruCache!!.put(key, bitmap)
    }

    operator fun get(key: String?): Bitmap? {
        return if (mLruCache == null) null else mLruCache!!.get(key)
    }

    companion object {

        var mLruCache: LruCache<String, Bitmap>? = null// 硬引用缓存

        val instance: MemCache
            get() = SingletonHolder.INSTANCE
    }
}